access/api.go

```diff
@@ -0,0 +1,136 @@
+package access
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/safing/portbase/api"
+	"github.com/safing/portbase/database/record"
+	"github.com/safing/spn/access/account"
+)
+
+func registerAPIEndpoints() error {
+	if err := api.RegisterEndpoint(api.Endpoint{
+		Path:        `spn/account/login`,
+		Read:        api.PermitAdmin,
+		BelongsTo:   module,
+		HandlerFunc: handleLogin,
+		Name:        "SPN Login",
+		Description: "Log into your SPN account.",
+	}); err != nil {
+		return err
+	}
+
+	if err := api.RegisterEndpoint(api.Endpoint{
+		Path:        `spn/account/logout`,
+		Read:        api.PermitAdmin,
+		BelongsTo:   module,
+		ActionFunc:  handleLogout,
+		Name:        "SPN Logout",
+		Description: "Logout from your SPN account.",
+		Parameters: []api.Parameter{
+			{
+				Method:      "GET",
+				Field:       "purge",
+				Value:       "",
+				Description: "If set, account data is purged. Otherwise, the username and device ID are kept in order to log into the same device when logging in with the same user again.",
+			},
+		},
+	}); err != nil {
+		return err
+	}
+
+	if err := api.RegisterEndpoint(api.Endpoint{
+		Path:        `spn/account/user/profile`,
+		Read:        api.PermitUser,
+		BelongsTo:   module,
+		RecordFunc:  handleGetUserProfile,
+		Name:        "SPN User Profile",
+		Description: "Get the user profile of the logged in SPN account.",
+		Parameters: []api.Parameter{
+			{
+				Method:      "GET",
+				Field:       "refresh",
+				Value:       "",
+				Description: "If set, the user profile is freshly fetched from the account server.",
+			},
+		},
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func handleLogin(w http.ResponseWriter, r *http.Request) {
+	// Check if we are already authenticated.
+	user, err := GetUser()
+	if err == nil && user.State != account.UserStateNone {
+		http.Error(
+			w,
+			fmt.Sprintf("Already logged in as %s (Device: %s)", user.Username, user.Device.Name),
+			http.StatusConflict,
+		)
+		return
+	}
+
+	// Get username and password.
+	username, password, ok := r.BasicAuth()
+	// Request, if omitted.
+	if !ok || username == "" || password == "" {
+		w.Header().Set("WWW-Authenticate", "Basic realm=SPN Login")
+		http.Error(w, "Login with your SPN account.", http.StatusUnauthorized)
+		return
+	}
+
+	// Process login.
+	user, code, err := login(module.Ctx, username, password)
+	if err != nil {
```

Comment:
use a sub-context with a deadline here because otherwise the portmaster might be blocked waiting for an unresponsive ticket agent